### PR TITLE
Create a PDF file from Markdown documents (md2pdf)

### PR DESCRIPTION
**Pull request by michelou**
WIP

The  [`pandoc`](https://github.com/michelou/dotty/tree/scala3-pandoc/pandoc) subproject make it possible to create a PDF file (e.g. [`scala3_reference.pdf`](https://github.com/michelou/dotty-examples/blob/master/docs/scala3_reference.pdf)) from the Markdown documents of the Scala 3 documentation (reference, internals, usage).

Todo:
- fix arguments passed to the Docker container
- test (and improve) workflow file [`Dockerfile`](https://github.com/michelou/dotty/blob/scala3-pandoc/pandoc/Dockerfile) to run CI job on GitHub Actions.

Related to PR #10767, #10826, #10860, #10875, #10953, #11016, #11158 and #11235.

### CODE REVIEW

**Comment by SethTisue** [comment]
I don't think there's anything here I can meaningfully review (I don't know Docker), but I'm delighted to see it happening.

**Comment by michelou** [comment]
> 
> I don't think there's anything here I can meaningfully review (I don't know Docker), but I'm delighted to see it happening.

That's fair. My primary intent was to let you know things are moving forward. I've removed your name from the reviewers.

File: pandoc/data/reference.md [hunk 1]
```diff
@@ -0,0 +1,566 @@
+---
+layout: doc-page
+title: "Scala 3 Reference"
+titlegraphic: ../out/pandoc/images/scala-spiral.png
+author: Martin Odersky and al.
```
Comment (liufengyun) [inline pandoc/data/reference.md:5]:
```suggestion
author: Martin Odersky et al.
```

File: pandoc/docs/PROJECT.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
+# <span id="top">More about the project</span> <span style="size:25%;"><a href="../README.md" title="Back to README">↩</a></span>
+<!-- created by mics (https://github.com/michelou/) on December 2020 -->
+
+This project is located at the root of the [`lampepfl/dotty`][github_dotty] repository, at the same level as the [`Scala3doc`][github_scala3doc] project.
+
+## <span id="structure">Directory structure</span>
```
Comment (liufengyun) [inline pandoc/docs/PROJECT.md:6]: It would be nice to merge this with `pandoc/README`, so that the reader has an overview on first read.

File: pandoc/docs/TEXLIVE.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
+# <span id="top">More about TeX Live</span> <span style="size:25%;"><a href="../README.md" title="Back to README">↩</a></span>
+<!-- created by mics (https://github.com/michelou/) on December 2020 -->
```
Comment (liufengyun) [inline pandoc/docs/TEXLIVE.md:2]: Is this document necessary given that we have the docker image?

File: pandoc/extras/build_image.bat [hunk 4]
```diff
@@ -0,0 +1,218 @@
+@rem ## Created by mics (https://github.com/michelou/) on December 2020.
```
Comment (liufengyun) [inline pandoc/extras/build_image.bat:1]: If image can be built by Linux/MacOS and published, then we don't need the script to build the image on Windows.

File: pandoc/README.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+- [Docker Desktop 3](https://docs.docker.com/get-docker/) <sup id="anchor_03">[[3]](#footnote_03)</sup> *([release notes](https://docs.docker.com/release-notes/))*
+
+> See document [`DOCKER.md`](docs/DOCKER.md) for further information, e.g. `Dockerfile` usage.
+
+## <span id="commands">Build commands</span> <sup style="font-size:60%;">[**&#9650;**](#top)</sup>
+
+We provide two commands for MacOS/Ubuntu as well as their equivalents for MS Windows :
+- for *interactive users* :<br/>[`build {<option>|<subcommand>}`](./build) (resp. [`build.bat`](./build.bat) on MS Windows)<br/><span style="font-size:80%;">(option `-help` displays the available options/subcommands).</span>
+- for *automated tasks* (primarily, but not exclusively) :<br/>[`md2pdf [<project>]`](./md2pdf) (resp. [`md2pdf.bat`](./md2pdf.bat) on MS Windows)<br/><span style="font-size:80%;">(default value for `<project>` : `reference`).</span>
```
Comment (liufengyun) [inline pandoc/README.md:48]: I've wondering, instead of having two commands `build` and `md2pdf`, can we have the same command for both local machines and cloud CI, assuming that both will use a published docker image?

**Review by liufengyun — COMMENTED** [review]
Very nice work 👍

File: pandoc/README.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+<p style="margin:0 0 1em 20px;">
+We run <a href="https://docs.docker.com/docker-for-windows/install/">Docker Desktop for Windows</a> on our development machine.<br/>
+Note that WSL 2 is required to interact with Docker on Windows from <a href="https://ubuntu.com/wsl">Ubuntu for WSL</a> (see documentation <a href="https://docs.docker.com/docker-for-windows/wsl/">Docker Desktop WSL 2 backend</a>).
+</p>
+
+***
+
+*[mics](https://github.com/michelou/)/January 2021* [**&#9650;**](#top "Back to top")
+<span id="bottom">&nbsp;</span>
```
[the remark this answers is not in the document]
Comment (michelou) [reply pandoc/README.md:81]: While your comment is justified for people working with 20+ inch screens, I and others I guess are used to read online docs with a 5.8 inch cell phone, in which case this tiny button is very practical. I let you decide.

File: pandoc/README.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+<!-- created by mics (https://github.com/michelou/) on December 2020 -->
+
+This project aims to generate a *single* PDF file from a collection of [Markdown] documents, for instance for the [*Scala 3 reference*][scala3_reference] documentation.
+
+Our solution is built on [Pandoc]; it gets its inputs from two directories and generates a PDF file into the output directory:
+
+<table style="margin-top:-10px; max-width:650px;">
+<tr>
+<td><b>Input directories</b></td>
+<td><code>dotty/docs/docs/&lt;project&gt;/</code><br/><code>dotty/pandoc/data/</code></td>
```
[the remark this answers is not in the document]
Comment (michelou) [reply pandoc/README.md:11]: Done.

File: pandoc/README.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+<td><code>*.md</code> files<br/><a href="https://pandoc.org/">Pandoc</a> data files</td>
+</tr>
+<tr>
+<td><b>Output directory</b></td>
+<td><code>dotty/out/pandoc/</code></td>
```
[the remark this answers is not in the document]
Comment (michelou) [reply pandoc/README.md:16]: Done,

File: pandoc/data/internals.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
+---
+layout: doc-page
+title: "Scala 3 Internals"
+titlegraphic: ../out/pandoc/images/scala-spiral.png
```
[the remark this answers is not in the document]
Comment (michelou) [reply pandoc/data/internals.md:4]: [`md2pdf` ](https://github.com/michelou/dotty/blob/scala3-pandoc/pandoc/md2pdf) starts with a preprocessing phase which transforms :
- cross-references found in the Markdown files (e.g. [`reference`](https://github.com/lampepfl/dotty/tree/master/docs/docs/reference)) to named anchors
- references to image files stored in `data/images/` (e.g. [`scala-spiral.png`](https://github.com/michelou/dotty/blob/scala3-pandoc/pandoc/data/images/scala-spiral.png)) to relative paths.

In short all references not starting with `https://` are transformed before calling the [`pandoc`](https://pandoc.org/MANUAL.html) command line tool. In the output directory `out/pandoc/`  both directories `images/` and `src_managed/` are expected to be at the same level : 
- `out/pandoc/images/`
- `out/pandoc/src_managed/`
- `out/scala3_reference.pdf`

File: pandoc/data/internals.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+author: Martin Odersky and al.
```
[the remark this answers is not in the document]
Comment (michelou) [reply pandoc/data/internals.md:5]: Done.

File: pandoc/data/reference.md (hunk 1 above)
Comment (michelou) [reply pandoc/data/reference.md:5]: Done.

File: pandoc/docs/PROJECT.md (hunk 2 above)
Comment (michelou) [reply pandoc/docs/PROJECT.md:6]: Done.

File: pandoc/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (liufengyun) [reply pandoc/README.md:81]: Fair enough, it's fine to keep it.

File: pandoc/README.md (hunk 5 above)
Comment (michelou) [reply pandoc/README.md:48]: Your suggestion is right.

The existence of both commands `build` and `md2pdf` stems in part from my own background as a long-standing developer. Let me try to advocate the case (*for fun*) : 
- as a preamble the existance of both scripts is not directly linked to Docker.
- [`build`](https://github.com/michelou/dotty/blob/scala3-pandoc/pandoc/build) is a user-friendly wrapper script; it calls `md2pdf` but it also
   - provides some options (e.g. `-verbose` and `-debug`) and some subcommands (e.g. `clean` and `view` on MS Windows); I typically run command `build -timer -verbose clean view` to generate file [`scala3_reference.pdf`](https://github.com/michelou/dotty-examples/blob/master/docs/scala3_reference.pdf) on my Win10 laptop.
   - provides conditional compilation; the output file (e.g. [`scala3_reference.pdf`](https://github.com/michelou/dotty-examples/blob/master/docs/scala3_reference.pdf)) is generated depending on input file timestamps.
-  [`md2pdf`](https://github.com/michelou/dotty/blob/scala3-pandoc/pandoc/md2pdf) is the minimal script for running [`pandoc`](https://pandoc.org/MANUAL.html) with the appropriate settings.

In the context of the Dotty project, we can indeed have *only one*  shell script :  [`md2pdf`](https://github.com/michelou/dotty/blob/scala3-pandoc/pandoc/md2pdf) (and may be [`md2pdf.bat`](https://github.com/michelou/dotty/blob/scala3-pandoc/pandoc/md2pdf.bat)).

**NB,** I could even omit several ~~type checks~~ environment checks in`md2pdf` assuming users can help themselves in case of errors.

File: pandoc/extras/build_image.bat (hunk 4 above)
Comment (michelou) [reply pandoc/extras/build_image.bat:1]: My wish here is to offer different ways to developers to generate a PDF file from the Markdown files *before* they publish their changes (more details per email).

**Comment by bishabosha** [comment]
Hi @michelou do you have any time to continue working on this?

**Comment by anatoliykmetyuk** [comment]
@michelou what's the status on this one? Are you still working on it? Do you need someone to review it?

**Comment by michelou** [comment]
> 
> 
> @michelou what's the status on this one? Are you still working on it? Do you need someone to review it?

Short status is as follows : 
1. I regularly update the **locally** generated PDF file [`scala3_reference.pdf`](https://github.com/michelou/dotty-examples/blob/master/docs/scala3_reference.pdf) (*last update: July 23, 2021*) on my GitHub repository [`dotty-examples`](https://github.com/michelou/dotty-examples). 
2. I made no progress with the generation of the PDF file in a **remote** Docker container because I failed to upload the Docker image (_NB._ my internet connection is very limited, i.e. download ~10 Mbps / upload ~1 Mbps).

In resume I need a co-worker ready to help me with **step 2**.

**Comment by griggt** [comment]
@michelou I think I can help with getting the PDF generation working in Docker / CI.  I have a branch from quite a while ago where I had this working....once I find it are you ok with me pushing the changes to this PR?

**Comment by michelou** [comment]
@anatoliykmetyuk @liufengyun @griggt Today I updated my *locally* generated PDF file [`scala3_reference.pdf`](https://github.com/michelou/dotty-examples/blob/master/docs/scala3_reference.pdf) .

Next steps :
- open one (or two) PR to integrate my changes to the Markdown files.
- give another try with the PDF generation built upon a Docker image.

**Comment by michelou** [comment]
Here is an actualized version of the **locally** generated PDF file [`scala3_reference.pdf`](https://github.com/michelou/dotty-examples/blob/master/docs/scala3_reference.pdf).

@griggt I finally have to bring myself to abandon my attempts with Docker. In short your help is welcome !
As explained in my previous messages all my work is contained in directory [`DOTTY_PROJECT_ROOT/pandoc/`](https://github.com/michelou/dotty/tree/scala3-pandoc/pandoc). If you agree we can discuss the next steps offline (aka. via email).

**Comment by michelou** [comment]
I just updated my GitHub project to (locally) generate the **actualized** PDF file [`scala3_reference.pdf`](https://github.com/michelou/dotty-examples/blob/master/docs/scala3_reference.pdf).
Otherwise, no progress on the Docker-side.

I plan to look at the recently added pages and submit a few additions/corrections.

**Comment by Gedochao** [comment]
It's been 5 years since the last push in this PR, I think it's safe to close it at this stage.
If someone is still working on this, feel free to open a new PR and link back here.
Either way, closing this.
